### ISSUE TITLE
Investigate means of authenticating a node ID when adding to `NodeGraph`

### ISSUE THREAD

**Issue by joshuakarp** [MatrixAI/Polykey#322]
### Specification
With #344 more or less solved we need a way of authenticating when adding forward connections. this includes authenticating the old connections  when a bucket fills up. In both cases we can so a `pingNode` to check if a node is alive.  However the current implementation of `pingNode` is insufficient for the job as it only checks if a node is alive.

`NodeManager.pingNode` needs to be updated to match the following criteria;
1. It needs to establish a proxy connection to the target. if a connection can be established then the target is considered alive. if the connection is established then we can consider it authenticated as well.
2. It needs to take an optional address information for checking if direct connections without the nat bashing proceedure can be done. This will be useful for ICE

The `NodeManager.setNode` needs to be updated to do the following.
1. when a bucket is full the oldest node needs to be verified with `pingNode`. This is already done, just need to update ping for this. This will need to concurrently check alpha of the oldest connections.
2. Needs an optional force parameter to garentee adding the node.

`NodeConnectionManager` needs to be updated to;
1. takes `NodeManager` as part of start.
2. establishing a connection should trigger `NM.qeueSetNode`

`NodeManager` needs to be updated to;
1. startStop asyinc init pattern.
2. implement a async queue for adding nodes to the nodeGraph. 

After that we need to make sure all instance where we need to add a node just calls `NodeManager.setNode`. Where this is done still needs to be speced out but this may be outside the scope of this issue. Create a new issue for this?

### Additional context

- another very relevant issue: #226
- Related #344 
- Related #326 
- Related #148 

### Tasks
- [x] update `NodeManager.pingNode` to match the above specification.
- [x] update `NodeManager.setNode` to match the above specification.
- [x] implement setnode queue. this is similar to the queue used by discovery. 
- [x] seperate pingNode out of setNode
- [x] concurrently check alpha old nodes if a bucket is full when setting node.
- [x] convert NCM and NM to startstop, they rely on each other now with NCM calling NM.qeueSetNode.
- [x] establishing a NodeConnection should trigger setting a node in the nodeGraph by callong NM.qeueSetNode.
- [x] Creating a new connection should not be blocked by setting node.
- [x] Integrate into `pk nodes set` (with `--no-ping` and `--force`) and `pk nodes ping`

**Comment by tegefaulkes** [comment]
There are 4 cases where we add a node to the node graph. 
1. manually 
2. ping a node
3. connect to a node in any way
4. receive a connection from a node

Manually adding a node should ping the node before adding it to the `NodeGraph`. In fact I think this is functionally identical to pinging a node in outcome except that manually adding the node should guarantee that the node is added to the graph by dropping an older node.

Pinging a node will add the node to the graph but this should be done in the CLI ping command's GRPC handler. having the `NodeManager.pingNode` is called to verify a node's 'aliveness' during an `setNode` so we can't have that attempting to set a node.

Adding a node after connecting to it can be handled by the nodeConnectionManager in the `with` context functions? or possible triggered by the forward proxy? need to explore this more. it would be nice to have this handled neatly in one place. 

receiving the connection will be verified by the reverse proxy and then added to the nodeGraph. this is addressed in #344

**Comment by tegefaulkes** [comment]
WIth the changes in commit 7cc74d059f94739591307d9543fa73f999d3345f I added a check for if a node was alive with `pingNode`. currently this is blocking and will timeout based on the proxies timeout which is usually 20 seconds. This will need to be fixed by letting us set a time out in the ping command. I think there was an issue planed for this? #353? but that's for the node connections, we will need a similar solution for the proxies stuff. New issue?

In any case we ill need to connect to a node and check it's nodeId and certificates data. maybe we can create a new agent-agent RPC for getting a remote node's certificates and nodeId? isn't this already what `GRPCClientAgent.nodesChainDataGet` does? we can verify the node id and MTLS on the client side as well.

**Comment by CMCDragonkai** [comment]
Yea the closest issue is #353, by creating a timeout for `NodeConnectionManager`.

I think however the proxies `openConnectionForward` and `openConnectionReverse` should also have timeouts that override the default timeout. This is already supported with the `timer` parameter.

Also relevant is #297 as background theory into the future.

**Comment by CMCDragonkai** [comment]
Furthermore solving #353 would just reduce the amount of time waiting for a ping. To really solve the problem of blocking, is to remove blocking entirely. That is we wouldn't want to block a request to contact a Node or do operations against another Node due to pinging old nodes in order to maintain bucket limit.

I suggest that this logic should be non-blocking, and to do so, we can either do non-blocking in-memory by queuing up a job using JS's own event loop, or by queuing it up into the DB and run this as part of a background work as suggested by the longer-kademlia trie paper at the end.

**Comment by CMCDragonkai** [comment]
Suggest creating a new issue for this non-blocking problem as it relates to #353.

**Comment by tegefaulkes** [comment]
Ah I just remembered that the feature added in 7cc74d059f94739591307d9543fa73f999d3345f just checks that the old node is still alive. as mentioned above authenticating it still needs to be done at this stage and that's only really needed to make sure that the node hasn't changed identities. it should've been authenticated before it was added to the graph. 

For adding a new node I think the expectation is that the authentication is done when connecting to it just before adding it to the nodeGraph. We could use the same method we use to authenticate the old nodes but we would be doubling up the action of connecting to the node since the action of adding a node is triggered by a connection to a node in the first place.

As to how the node is authenticated we would need to obtain the connection info from the newly combined `Proxy` class. Based on this comment https://github.com/MatrixAI/js-polykey/issues/344#issuecomment-1064806472 we will add a callback to the `Proxy` class that will cause a connection event in the `EventBus` system. This will trigger adding the node to the node graph. Assuming that the proxy authenticates the connection before triggering the event then the only information the event needs is a tuple of `(remoteNodeId, remoteHost, remotePort)`.

**Comment by tegefaulkes** [comment]
Ignore this, moved old spec back.

**Comment by CMCDragonkai** [comment]
I think the proxy open connection already does authentication by checking the certificate. Therefore the needing to authenticate a node would only be when one is explicitly adding a node to the node graph.

Any authentication would need to be done prior to adding the node to the node graph.

One question is whether we should allow users to explicitly add nodes to the node manager/graph straight from the command line and therefore not have to be authenticated ahead of time. I think we should allow this as this is would primarily be used for special situations and debugging.

The node manager's add node method can schedule an authentication by scheduling a ping node.

I think the ping node should not just establish a proxy connection but also check the certificates which I believe the proxy does already. But you should check @tegefaulkes

**Comment by CMCDragonkai** [comment]
We should re-use the ping node as the basis of "authenticating a node". It would then be used when refreshing buckets and dealing with filled buckets.

We could add a boolean flag during the add node method which controls whether to authenticate the node prior to adding.

**Comment by tegefaulkes** [comment]
Right now the `setNode` assumes the node was already authenticated prior to calling `setNode` since this will be triggered by an already authenticated connection by the `Proxy`.

**Comment by tegefaulkes** [comment]
Hmm.. the spec I created here relates more to issue #344  but it still relates this one partly. The scope this issue relates more to authenticating and adding nodes when connecting **TO** a node. Whereas 344 relates to authenticating and adding a node when receiving a connection **FROM** a node.

I hoped that I could handle adding nodes for `ForwardConnection`s the same way as the `ReverseConnection`s. But if we automatically try to add the old node when we establish the connection for a ping then that could lead to issues or it's too side effect-ful.

So adding nodes based on reverse connections can be automatic but we may have to handle forward connections separately. 

In any case, this Issue should focus more on the authenticating the forward connections and pinging the old nodes when we need to add a new node to a full bucket.

**Comment by tegefaulkes** [comment]
I reverted to the old spec. This needs to be spec-ed out focusing on authenticating and adding nodes on forward connections. the `pingNode` should also authenticate the node it's pinging. 

So maybe like mentioned above, we can just use the `pingNode` to authenticate the node as well. we can ping the node when adding the node to the node with `NodeManager.setNode`. add a flag to `NodeManager.setNode` to skip the 'authenticating` ping for the new node, this will be needed for manually adding nodes and the adding nodes on reverse connection.

The `pingNode` function I think will need to become a agent-agent RPC, utilize the `NodeConnectionManager`'s connections. it will need to return true if a `NodeConnection` can be established, was authenticated and was the expected NodeId. We also don't want to attempt a kademlia `FIND_NODE` operation when trying to ping since we are trying to verify the address. Is that possible with the NodeConnectionManager? 

I'll have to review code and spec this futher.

**Comment by tegefaulkes** [comment]
Old spec
---
### Specification

We currently have some instances when adding nodes to our `NodeGraph` where we attempt to perform a connection to the node before adding it to the `NodeGraph` buckets database.

Previously, this was intended as a primitive means of ensuring that only node IDs that corresponded with valid keynodes were added to one's own `NodeGraph` database (i.e. keeping the state "clean"). We could assume that a node ID was valid if we connected to it and performed the certificate validation with respect to its expected node ID.

This problem relates to mitigating the impact malicious actors/keynodes have on the wider Polykey network, and keeping the network as "clean" as possible. There needs to be further discussion about how this will be considered with Polykey, with this `NodeGraph` issue being one part of that discussion.

### Additional context

- I've removed two instances of this in #310 here https://github.com/MatrixAI/js-polykey/pull/310/commits/6cfc97691d16cb94f2cfb2f49b289e14a62c8873
- another very relevant issue: #226

### Tasks

1. Find all instances where we add nodes to our own `NodeGraph`. (There may not be any more.)
2. Remove these in favour of some other mechanism.

**Comment by tegefaulkes** [comment]
I feel like I keep editing over issues that describe problems to turn them into design specifications. I feel that we should keep the issues that describe problems as they are. When we find a solution to the problem we should create a new design issue specifying the solution and design spec. Basically I think we should have separation between problem definitions and problem solutions. 

```
problems    solutions       PRs
A -------------->D --------> F
B ---------------^           ^
C -------------->E ----------|
```

It's just a though I had.
@CMCDragonkai

**Comment by tegefaulkes** [comment]
tests relating to new `pingNode` spec
- Can find the node to ping with kademlia find_node.
- pings the node directly with the optional host/port parameter
- ping fails if the node fails to authenticate.
- ping fails if target authenticates but doesn't match the expected NodeId.
- ping attempts a NAT punchthrough.
- ping fails if we fail to connect to the target.
- ping success case.

**Comment by CMCDragonkai** [comment]
> I feel like I keep editing over issues that describe problems to turn them into design specifications. I feel that we should keep the issues that describe problems as they are. When we find a solution to the problem we should create a new design issue specifying the solution and design spec. Basically I think we should have separation between problem definitions and problem solutions.
> 
> ```
> problems    solutions       PRs
> A -------------->D --------> F
> B ---------------^           ^
> C -------------->E ----------|
> ```
> 
> It's just a though I had. @CMCDragonkai

It's fine to edit issues, in fact they should be edited. We don't want to create too many issues, as old information becomes out-dated. So editing is just a natural process of pruning information and updating information. Makes it easier to do the R&D report later.

**Comment by tegefaulkes** [comment]
Looking into how to implement the `pingNode` properly. It seems that establishing a `NodeConnection` in the `NodeConnectionManager` is functionally identical to our needs for pinging a node. All we really need to do is to expand the NodeConnectionManager to allow us to specify the host and Port when creating a connection.

So we can add a `pingNode` method to the `NodeConnectionManager` and have it use a modified `createConnection` to allow us to specifiy the target host and port. We will need to make sure that the `NodeConnection` authenticates the connection which it likely already does. and make sure the the authentecated Node matches the target node. 

May as well add optional connection timeouts here as well.

Do you think this is a good course of action @CMCDragonkai ?

**Comment by tegefaulkes** [comment]
Wild question, How do we actually get the host and port of the client connecting to the forward proxy? `nodeConnection.port` and `client.port` seems to return the proxy port.

**Comment by tegefaulkes** [comment]
`pingNode` implementation has been fixed up. now to focus on the `setNode` part of this.

As for making all of this non-blocking. I think that should be speced out in a separate issue.

**Comment by tegefaulkes** [comment]
Ok, it seems that establishing a `NodeConnection` to do a ping is fine but is a bit too expensive for our use. Ideally we will just establish a hole-punch connection using just the `Proxy`. This however depends on some refactoring work done for how hole-punching and related messaging. 

In any case the ping method should be refactored to do mostly the same thing but establish a connection using the `Proxy` in stead of the `NodeConnectionManager` this means we will need to handle the hole-punch relay message within the ping command. But I think the whole `hole-punch` process needs to be refactored and separated outside of the `NodeConnection` as it's own process.

In the mean time I could focus on the `setNode` implementation for now. But with the since authenticating nodes in the background still needs to be specced out then maybe this issue should be put on hold?

**Comment by tegefaulkes** [comment]
Right now this is blocked by some changes relating to how the `Proxy` and `NodeConnectionManager` starts a `NodeConnection` through a NAT. I will need to create a new issue for this. this will also relates to #365.

**Comment by CMCDragonkai** [comment]
For this issue, we have to establish:

1. Opening a connection already verifies the `NodeId` (and thus does the authentication) due to MTLS.
2. Setting a node in the `NodeGraph` due to opening a connection is therefore already correct, as the node entry is authenticated.
3. This issue has to tackle 3 cases (where you have to "authenticate" the node entry):
   1. Adding a node manually from the CLI, and using `--force` flag
   2. Getting node entries from another node due to `getClosestNode` call
   3. Refreshing the `NodeGraph` in order to prune existing entries

**Comment by CMCDragonkai** [comment]
For `3.i` we can have:

```
# set a node, but if the bucket is full, it goes into the set node policy
pk nodes set <nodeID> <nodeAddress>

# set a node, and if the bucket is full, drop the oldest node in the bucket (don't even use the concurrency parameter for the oldest)
pk nodes set <nodeID> <nodeAddress> --force

# set a node without checking if that node is live
pk nodes set <nodeID> <nodeAddress> --no-ping

# don't check old nodes, and don't check the new node
pk nodes set <nodeID> <nodeAddress> --force --no-ping
```

At the methods, these can be `force=true/false` and `ping=true/false`.

Default is `force=false` and `ping=true`.

**Comment by CMCDragonkai** [comment]
For `3.ii` this is actually more generic. Because it's more about whenever we are creating a node connection we want that entry to be part of the node graph.

To do this, some major changes will be required.

### Proxy

1. We will need a `composeTimeoutTime` as a TTL for `ConnectionForward` and `ConnectionReverse` (use `Connection` abstract class) for proxy connections that have not yet been composed. So we can set a `setTimeout` call that when executed checks if it is being composed (or in the middle of composing), and if not, destroy the connection. This will allow proxy connections to be created without having to worry about tracking them to destroy, since "un-composed" connections will eventually be destroyed.
2. Once a connection is composed, it is always composed.
3. Persistent proxy connections are by themselves not possible for maintaining NAT reachability, it will have to be done at the `NodeConnection` level. See #365.
4. The `openConnectionForward` and `openConnectionReverse` should be memoized for live started connections. They should always give you back a pre-existing connection. Because these are concurrent, they should also have creation-locking and follow the object-map pattern (concurrent calls should coalesce). A combination of memoization and compose TTL enables a "time-based cache" of proxy connections.

### NodeManager, NodeConnectionManager

1. The `setNode` method has to deal with garbage collecting old nodes in a full bucket in case the `force = false`.
2. The `setNode` may require to ping the node in order to "authenticate", but this can be an expectation of the node parameter itself, forcing higher level ping node like `pk nodes set` to do `pingNode` explicitly prior to `setNode`.
3. When garbage collecting old nodes, it can contact multiple oldest nodes. This may be the concurrency constant.
4. Contacting old nodes means using `NodeManager.pingNode` which will end up creating proxy connections and rely on the compose TTL mechanism to avoid having to destroy the proxy connections immediately upon completion. Contacting old node SHOULD NOT lead to `setNode` again, this would cause a loop.
5. NCM and NM both needs to be `StartStop` because they are going to be tightly coupled now, and both needs each other during the start/stop pattern. The NG is created separately and passed into both during `start`.
6. New node connections should trigger the node graph to contain the new node entry (but can be done in an eventually-consistent manner.)
6. Starting a new node connection should not be delayed by the pinging of the target node, and any garbage collection of old nodes. This means `setNode` has to be "queued".
7. A queue data structure managed by the `NodeManager` in a similar way to `Discovery` can be used to asynchronously process `setNode` calls. This could be done with `NodeManager.queueSetNode()` which is actually synchronous rather than asynchronous. Unlike `Discovery`, this queue would in-memory only, and not need to be persistent. However if #329 is done, then we could reuse the same code to the queuing, in which case the queue would end up being persistent.
8. During the creation on a new node connection, whether this is from `NCM.withNodeConn` or otherwise, this has to utilise `queueSetNode`. Such an operation should not ping the target node, because if the connection is successful, then the target node is authenticated. If the connection is unsuccessful, this should be cancelled. If successful, the set node functionality should then garbage collect old nodes in the background. This could mean:
   ```
   establishNodeConnection (nodeId) {
     await NodeConnection.createNodeConnection(nodeId);
     nodeManager.queueSetNode(nodeId);
   }
   ```
9. We cannot just hook the `queueSetNode`  into proxy forward connection establishment, because this would cause a loop when you just want to ping a node without also setting the node like during garbage collection of old nodes in a full bucket.

**Comment by tegefaulkes** [comment]
With the 6324bb33 commit I think this issue is pretty much addressed.

**Comment by emmacasolin** [comment]
Do the changes to `pingNode()` in this issue also address https://github.com/MatrixAI/js-polykey/issues/148?

**Comment by CMCDragonkai** [comment]
Maybe, remember ping node has to establish a proxy connection which always automatically validates the certificate which basically means any established proxy connection is by definition authenticated.

**Comment by emmacasolin** [comment]
> Maybe, remember ping node has to establish a proxy connection which always automatically validates the certificate which basically means any established proxy connection is by definition authenticated.

Ok, I'll add it into the issue description here. It may require further review to ensure it's being fully addressed, but it sounds like it's definitely related.

**Comment by emmacasolin** [comment]
> * [Investigate means of authenticating a node ID when adding to `NodeGraph` #322](https://github.com/MatrixAI/js-polykey/issues/322) - Is done. Need to check that in all cases we verify a connection to a node before adding it to the node graph. I know that `NodeConnectionManager.syncNodeGraph` isn't doing it already but that needs some extra logic to do it efficiently. https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1098666122

I think the crux of this is that we want to connect to a node before we add it to the node graph, since this way we can check if it's valid before doing so. The problem for `syncNodeGraph` though is that we need to add multiple nodes at once, and attempting to create a connection to every one of them is inefficient, especially if any of them are offline (which isn't all that unlikely).

I think the options here are:
a) Skip creating a connection to nodes that are added by `syncNodeGraph`
- Since these node id -> address mappings are coming from a seed node (which should be an agent we can trust) it's probably a reasonable assumption that these mappings are correct (or were at least recently correct)
- If a mapping does turn out to be incorrect, it will eventually be removed from the node graph as new nodes get added in, so we will *eventually* have only correct mappings in our node graph

b) Attempt to create a connection to all of the nodes at the same time, similar to what's done in `NodeManager.garbageCollectOldNode`
- We still need to wait, but the amount of time we need to wait is constant regardless of how many nodes we have to check

c) Create another asynchronous queue for `pingNode`
- Going the asynchronous route means that we won't be blocked when calling `syncNodeGraph`
- The problem with this approach though is that this complicates the process of setting the nodes into the node graph - the ping queue and set queue would need to be linked somehow, such that a successful ping causes the pinged node to be added to the `setNode` queue to be added to the node graph
- This solution also runs the risk of nodes that were successfully pinged and added to the `setNode` queue subsequently going offline by the time they get added to the node graph, however, this is the same risk as for any other node in the `setNode` queue

**Comment by emmacasolin** [comment]
There are still a few places where we're not verifying a connection to a node before adding it to the node graph:

- `src/bin/nodes/CommandAdd.ts`
  - This is the command for setting a node into the node graph
  - It has not been updated with the `--force` and `--no-ping` options from https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1082638907
- `src/client/service/nodesAdd.ts`
  - This is the RPC for `CommandAdd`
  - It does not call `nodesPing` before calling `setNode`
  - It also has not been updated with the `--force` and `--no-ping` options, which would affect whether or not we need to call `nodesPing` for the node we're adding
- `src/nodes/NodeConnectionManager.ts`
  - `start`
    - When setting the seed nodes into the node graph we aren't connecting to them first
    - This may be ok though since seed nodes should always be online, however, there is a possibility that the seed node details added during the creation of the node could be incorrect so there's a potential security risk there I would imagine
  - `getClosestGlobalNodes`
    - Here we're calling `setNode` first, then `getConnection` - should this be the other way around?
    - Also we're calling `setNode` on the found nodes without connecting to them at all
  - `syncNodeGraph`
    - Already discussed https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1102102021

**Comment by emmacasolin** [comment]
> For `3.i` we can have:
> 
> ```
> # set a node, but if the bucket is full, it goes into the set node policy
> pk nodes set <nodeID> <nodeAddress>
> 
> # set a node, and if the bucket is full, drop the oldest node in the bucket (don't even use the concurrency parameter for the oldest)
> pk nodes set <nodeID> <nodeAddress> --force
> 
> # set a node without checking if that node is live
> pk nodes set <nodeID> <nodeAddress> --no-ping
> 
> # don't check old nodes, and don't check the new node
> pk nodes set <nodeID> <nodeAddress> --force --no-ping
> ```
> 
> At the methods, these can be `force=true/false` and `ping=true/false`.
> 
> Default is `force=false` and `ping=true`.

Since `-np` is already taken by `--node-path` should `--no-ping` just not have a shortened version? Or should we rename it?

**Comment by emmacasolin** [comment]
With https://github.com/MatrixAI/js-polykey/pull/326/commits/d3878482e6233e3a936b8cb8e7b946ff4e60afe0 and https://github.com/MatrixAI/js-polykey/pull/326/commits/752be04accc169d1e458ae940c7c585223582e6a `syncNodeGraph` is now non-blocking. Both `NodeManager.setNode` and `NodeConnectionManager.syncNodeGraph` utilise the same queue for setting new nodes into the node graph. In the future, this queue class can be generalised even further so that it can be used for other parts of the codebase that require a queue, e.g. Discovery.

**Comment by CMCDragonkai** [comment]
@tegefaulkes I believe the `pk nodes set` and `pk nodes ping` command still need to be updated to support this.

By default the `pk nodes set` should automatically ping the node. But with a `--no-ping` flag, it will not ping. But if using `--force` will also force it into the bucket.

**Comment by CMCDragonkai** [comment]
Using --no-ping is fine. Unless we use a different word to ping.

On 20 April 2022 10:00:02 am AEST, emmacasolin ***@***.***> wrote:
>> For `3.i` we can have:
>> 
>> ```
>> # set a node, but if the bucket is full, it goes into the set node policy
>> pk nodes set <nodeID> <nodeAddress>
>> 
>> # set a node, and if the bucket is full, drop the oldest node in the bucket (don't even use the concurrency parameter for the oldest)
>> pk nodes set <nodeID> <nodeAddress> --force
>> 
>> # set a node without checking if that node is live
>> pk nodes set <nodeID> <nodeAddress> --no-ping
>> 
>> # don't check old nodes, and don't check the new node
>> pk nodes set <nodeID> <nodeAddress> --force --no-ping
>> ```
>> 
>> At the methods, these can be `force=true/false` and `ping=true/false`.
>> 
>> Default is `force=false` and `ping=true`.
>
>Since `-np` is already taken by `--node-path` should `--no-ping` just not have a shortened version? Or should we rename it?
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/MatrixAI/js-polykey/issues/322#issuecomment-1103275027
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.
